nntrainer/src/neuralnet.cpp

```diff
@@ -542,39 +536,34 @@ int NeuralNetwork::isInitializable() {
     break;
   }
 
-  std::unordered_set<std::string> layer_name_set;
-
-  for (auto layer : layers) {
-    const std::string &name = layer->getName();
-    status = layer->checkValidation();
-    if (status != ML_ERROR_NONE) {
-      ml_loge("layer(%s) is not initializable", name.c_str());
-      return status;
-    }
-
-    if (layer_name_set.count(name)) {
-      ml_loge("layer(%s) name is duplicated", name.c_str());
-      return ML_ERROR_INVALID_PARAMETER;
-    }
-
-    layer_name_set.insert(name);
-  }
-
-  return status;
+  return ML_ERROR_NONE;
 }
 
 int NeuralNetwork::addLayer(std::shared_ptr<Layer> layer) {
   int status = ML_ERROR_NONE;
 
-  LayerType type = layer->getType();
-  if (type == LAYER_UNKNOWN)
-    return ML_ERROR_INVALID_PARAMETER;
-
   if (initialized) {
     return ML_ERROR_NOT_SUPPORTED;
   }
 
   ensureName(layer);
+
+  /** Validate the layer to be added */
+  status = layer->checkValidation();
```

Comment:
If we validate and add layer_names only inside `addLayer`, it should be the single source of writing the layer. I think it'd be best if we could restrict the write access for the `layers`.

For example, `layers.insert` is being used for some methods.

Reply:
Yes, insertion of layers would be inserted once we have graph object. That will have a single interface to add layer via the graph object. 
In the current code, as a hot fix, each `layer.insert()` is attached with `ensureName(layer)` for this purpose. Graph will solve all these issues. (I can update `addLayer` to include a position and then `addLayer` would insert at that position than at the end, but anyway graph will be coming next month, so skipping it.)

Reply:
Sounds fare enough